# Initial EDA

This notebook will serve to perform some initial EDA to get a better sensing about the data and its distribution. Will be using Fireducks (only for Linux :( ) because it's supposedly faster than pandas.

## Loading Dependencies

In [5]:
!pip install opendatasets

     ---------------------------------------- 0.0/82.7 kB ? eta -:--:--
     ---------------------------------------  81.9/82.7 kB ? eta -:--:--
     ---------------------------------------  81.9/82.7 kB ? eta -:--:--
     ---------------------------------------  81.9/82.7 kB ? eta -:--:--
     ---------------------------------------  81.9/82.7 kB ? eta -:--:--
     ---------------------------------------  81.9/82.7 kB ? eta -:--:--
     ---------------------------------------  81.9/82.7 kB ? eta -:--:--
     ---------------------------------------  81.9/82.7 kB ? eta -:--:--
     -------------------------------------- 82.7/82.7 kB 231.5 kB/s eta 0:00:00
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
   ---------------------------------------- 0.0/78.2 kB ? eta -:--:--
   ------------------------------------ --- 71.7/78.2 kB 3.8 MB/s eta 0:00:01
   ---------------------------------------- 78.2/78.2 kB 1.4 MB/s eta 0:00:00
  Created

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.

[notice] A new release of pip is available: 23.3.2 -> 24.2
[notice] To update, run: C:\Users\kengb\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [1]:
# import fireducks.pandas as pd
import pandas as pd
import opendatasets as od
import time, IPython
import numpy as np

# download the dataset
# od.download("https://www.kaggle.com/datasets/bulter22/airline-data/data")

In [2]:
carriers_path = "./airline-data/carriers.csv"
carrier_df = pd.read_csv(carriers_path)
carrier_df

Code                    Description
0        02Q                  Titan Airways
1        04Q             Tradewind Aviation
2        05Q            Comlux Aviation, AG
3        06Q  Master Top Linhas Aereas Ltd.
4        07Q            Flair Airlines Ltd.
...      ...                            ...
1486      ZW    Air Wisconsin Airlines Corp
1487      ZX                   Air Georgian
1488  ZX (1)                     Airbc Ltd.
1489      ZY         Atlantic Gulf Airlines
1490     ZYZ           Skyway Aviation Inc.

[1491 rows x 2 columns]

In [5]:
carrier_df[carrier_df.Code=='NW']

Code              Description
920   NW  Northwest Airlines Inc.

In [7]:
# need to process the shuffle file
pathfile='./airline-data/airline.csv.shuffle'       
reader = pd.read_csv(pathfile, chunksize=10**6, iterator=True, encoding='iso8859-1')       

def print_mem_usage(df):
    mem = df.memory_usage().sum() / 1024 ** 2
    print('The DataFrame takes up {:.3} MB'.format(mem))

train = pd.DataFrame()
start=time.time()  
for i, chunk in enumerate(reader): 
    train = pd.concat([train, chunk.sample(frac=.08, replace=False, random_state=213) ], axis=0)  
    if i % 20 == 0:
        print('Processing Chunk No. {}'.format(i))  
print('the program costs %.2f seconds'%(time.time() - start))
    
print_mem_usage(train)
print('train has {} rows and {} columns'.format(train.shape[0], train.shape[1]))

train.to_csv('airline.csv', index=False)

Processing Chunk No. 0
Processing Chunk No. 20
Processing Chunk No. 40
Processing Chunk No. 60
Processing Chunk No. 80
Processing Chunk No. 100
Processing Chunk No. 120
the program costs 1335.48 seconds
The DataFrame takes up 2.26e+03 MB
train has 9882798 rows and 29 columns


## Exploring Data

In [10]:
airline_path = "./airline.csv"
airline_df = pd.read_csv(airline_path, encoding='utf-8')
airline_df

ActualElapsedTime  AirTime  ArrDelay  ArrTime  CRSArrTime  \
0                    154.0    122.0      90.0   1850.0        1720   
1                    159.0    125.0      59.0   1703.0        1604   
2                      NaN     71.0       NaN      NaN        1140   
3                     70.0      NaN      80.0     20.0        2300   
4                    150.0    135.0       5.0   2030.0        2025   
...                    ...      ...       ...      ...         ...   
9882793              115.0     93.0       5.0    906.0           0   
9882794              137.0      NaN     -10.0    708.0         718   
9882795               76.0     61.0      -4.0    736.0         740   
9882796              157.0    124.0      18.0      3.0        2345   
9882797                NaN      NaN       NaN      NaN         740   

         CRSDepTime  CRSElapsedTime CancellationCode  Cancelled  CarrierDelay  \
0              1455           145.0              NaN          0           0.0   
1              1510           114.0              NaN          0           NaN   
2              1037            63.0              NaN          1           NaN   
3              2100            60.0              NaN          0           NaN   
4              1740           165.0              NaN          0           NaN   
...             ...             ...              ...        ...           ...   
9882793           0           119.0              NaN          0           NaN   
9882794         550           148.0              NaN          0           NaN   
9882795         620            80.0              NaN          0           NaN   
9882796        2015           150.0              NaN          0           0.0   
9882797         700            40.0              NaN          1           NaN   

         ...  Month  NASDelay  Origin  SecurityDelay TailNum  TaxiIn  TaxiOut  \
0        ...      6      23.0     ORD            0.0  N293AA     8.0     24.0   
1        ...      9       NaN     DTW            NaN  N8921E    19.0     15.0   
2        ...      2       NaN     CVG            NaN  N331DL     4.0     15.0   
3        ...      2       NaN     MDW            NaN     NaN     NaN      NaN   
4        ...      2       NaN     PHL            NaN  N512AU     3.0     12.0   
...      ...    ...       ...     ...            ...     ...     ...      ...   
9882793  ...     12       NaN     DFW            NaN  N579AA     9.0     13.0   
9882794  ...      8       NaN     MCO            NaN     NaN     NaN      NaN   
9882795  ...      5       NaN     ONT            NaN  N639@@     3.0     12.0   
9882796  ...     12       7.0     SEA            0.0  N916FR     6.0     27.0   
9882797  ...      1       NaN     AZO            NaN     NaN     NaN      NaN   

         UniqueCarrier  WeatherDelay  Year  
0                   AA           0.0  2006  
1                   NW           NaN  1997  
2                   DL           NaN  1995  
3               ML (1)           NaN  1991  
4                   US           NaN  1997  
...                ...           ...   ...  
9882793             AA           NaN  1997  
9882794             DL           NaN  1992  
9882795             WN           NaN  2001  
9882796             F9           0.0  2005  
9882797             NW           NaN  1990  

[9882798 rows x 29 columns]

In [16]:
years = list(airline_df.Year.unique())
years.sort()
years

[1987,
 1988,
 1989,
 1990,
 1991,
 1992,
 1993,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008]

In [ ]:
airline_df

In [11]:

airline_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9882798 entries, 0 to 9882797
Data columns (total 29 columns):
 #   Column             Dtype  
---  ------             -----  
 0   ActualElapsedTime  float64
 1   AirTime            float64
 2   ArrDelay           float64
 3   ArrTime            float64
 4   CRSArrTime         int64  
 5   CRSDepTime         int64  
 6   CRSElapsedTime     float64
 7   CancellationCode   object 
 8   Cancelled          int64  
 9   CarrierDelay       float64
 10  DayOfWeek          int64  
 11  DayofMonth         int64  
 12  DepDelay           float64
 13  DepTime            float64
 14  Dest               object 
 15  Distance           float64
 16  Diverted           int64  
 17  FlightNum          int64  
 18  LateAircraftDelay  float64
 19  Month              int64  
 20  NASDelay           float64
 21  Origin             object 
 22  SecurityDelay      float64
 23  TailNum            object 
 24  TaxiIn             float64
 25  TaxiOut           

In [10]:
airline_df.columns

Index(['ActualElapsedTime', 'AirTime', 'ArrDelay', 'ArrTime', 'CRSArrTime',
       'CRSDepTime', 'CRSElapsedTime', 'CancellationCode', 'Cancelled',
       'CarrierDelay', 'DayOfWeek', 'DayofMonth', 'DepDelay', 'DepTime',
       'Dest', 'Distance', 'Diverted', 'FlightNum', 'LateAircraftDelay',
       'Month', 'NASDelay', 'Origin', 'SecurityDelay', 'TailNum', 'TaxiIn',
       'TaxiOut', 'UniqueCarrier', 'WeatherDelay', 'Year'],
      dtype='object')

## Checking Cancelled FLights

In [19]:
cancelled_airlines = airline_df[~airline_df["CancellationCode"].isnull()]
cancelled_airlines.shape

(58620, 29)

In [20]:
cancelled_airlines

ActualElapsedTime  AirTime  ArrDelay  ArrTime  CRSArrTime  \
73                     NaN      NaN       NaN      NaN        1748   
242                    NaN      NaN       NaN      NaN        1021   
295                    NaN      NaN       NaN      NaN         827   
336                    NaN      NaN       NaN      NaN        1711   
727                    NaN      NaN       NaN      NaN        1137   
...                    ...      ...       ...      ...         ...   
9882126                NaN      NaN       NaN      NaN        1120   
9882462                NaN      NaN       NaN      NaN         830   
9882674                NaN      NaN       NaN      NaN        2335   
9882704                NaN      NaN       NaN      NaN         908   
9882742                NaN      NaN       NaN      NaN        2017   

         CRSDepTime  CRSElapsedTime CancellationCode  Cancelled  CarrierDelay  \
73             1544           124.0                C          1           0.0   
242             925            56.0                A          1           0.0   
295             610           257.0                A          1           NaN   
336            1335           156.0                C          1           0.0   
727            1055            42.0                A          1           0.0   
...             ...             ...              ...        ...           ...   
9882126         840           220.0                B          1           0.0   
9882462         630           120.0                B          1           0.0   
9882674        2230            65.0                C          1           0.0   
9882704         745            83.0                A          1           0.0   
9882742        1730           167.0                A          1           0.0   

         ...  Month  NASDelay  Origin  SecurityDelay TailNum  TaxiIn  TaxiOut  \
73       ...      5       0.0     IAH            0.0       0     0.0      0.0   
242      ...      2       0.0     RDU            0.0       0     0.0      0.0   
295      ...      5       NaN     ORD            NaN     NaN     NaN      NaN   
336      ...      8       0.0     IAH            0.0       0     0.0      0.0   
727      ...      1       0.0     COS            0.0  000000     0.0      0.0   
...      ...    ...       ...     ...            ...     ...     ...      ...   
9882126  ...     12       0.0     PVD            0.0       0     0.0      0.0   
9882462  ...     11       0.0     GSP            0.0  N940CA     0.0      0.0   
9882674  ...     10       0.0     JFK            0.0       0     0.0      0.0   
9882704  ...     10       0.0     PIT            0.0       0     0.0      0.0   
9882742  ...      8       0.0     LGA            0.0  N651DL     0.0      0.0   

         UniqueCarrier  WeatherDelay  Year  
73                  XE           0.0  2006  
242                 US           0.0  2005  
295                 UA           NaN  2008  
336                 CO           0.0  2004  
727                 UA           0.0  2004  
...                ...           ...   ...  
9882126             WN           0.0  2005  
9882462             OH           0.0  2006  
9882674             MQ           0.0  2007  
9882704             US           0.0  2004  
9882742             DL           0.0  2004  

[58620 rows x 29 columns]

In [22]:
cancelled_airlines.info()

<class 'pandas.core.frame.DataFrame'>
Index: 58620 entries, 73 to 9882742
Data columns (total 29 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ActualElapsedTime  0 non-null      float64
 1   AirTime            0 non-null      float64
 2   ArrDelay           0 non-null      float64
 3   ArrTime            0 non-null      float64
 4   CRSArrTime         58620 non-null  int64  
 5   CRSDepTime         58620 non-null  int64  
 6   CRSElapsedTime     58576 non-null  float64
 7   CancellationCode   58620 non-null  object 
 8   Cancelled          58620 non-null  int64  
 9   CarrierDelay       47661 non-null  float64
 10  DayOfWeek          58620 non-null  int64  
 11  DayofMonth         58620 non-null  int64  
 12  DepDelay           105 non-null    float64
 13  DepTime            105 non-null    float64
 14  Dest               58620 non-null  object 
 15  Distance           58620 non-null  float64
 16  Diverted           58620

In [25]:
len(airline_df[airline_df["Cancelled"]==1])

184011

In [27]:
# saving df first cause too many records and the main dataset is too big to work with properly
airline_df[airline_df["Cancelled"]==1].to_csv("cancelled_flights.csv", index_label=False)

In [35]:
top_cancellations = cancelled_airlines.UniqueCarrier.value_counts().index.to_list()
carrier_df[carrier_df['Code'].isin(top_cancellations)]

Code                                        Description
84     9E                             Pinnacle Airlines Inc.
100    AA                             American Airlines Inc.
206    AQ                                Aloha Airlines Inc.
221    AS                               Alaska Airlines Inc.
263    B6                                    JetBlue Airways
377    CO                         Continental Air Lines Inc.
434    DH                                   Independence Air
441    DL                               Delta Air Lines Inc.
505    EV                        Atlantic Southeast Airlines
517    F9                             Frontier Airlines Inc.
537    FL                        AirTran Airways Corporation
619    HA                             Hawaiian Airlines Inc.
642    HP  America West Airlines Inc. (Merged with US Air...
848    MQ                       American Eagle Airlines Inc.
920    NW                            Northwest Airlines Inc.
936    OH                                        Comair Inc.
949    OO                              Skywest Airlines Inc.
1292   TZ                             ATA Airlines d/b/a ATA
1297   UA                              United Air Lines Inc.
1308   US  US Airways Inc. (Merged with America West 9/05...
1388   WN                             Southwest Airlines Co.
1447   XE                           Expressjet Airlines Inc.
1465   YV                                 Mesa Airlines Inc.

In [37]:
cancelled_airlines = pd.merge(cancelled_airlines, carrier_df, how='left', left_on='UniqueCarrier', right_on='Code')
cancelled_airlines.drop(columns=['Code'])

In [46]:
cancelled_airlines.columns

Index(['ActualElapsedTime', 'AirTime', 'ArrDelay', 'ArrTime', 'CRSArrTime',
       'CRSDepTime', 'CRSElapsedTime', 'CancellationCode', 'Cancelled',
       'CarrierDelay', 'DayOfWeek', 'DayofMonth', 'DepDelay', 'DepTime',
       'Dest', 'Distance', 'Diverted', 'FlightNum', 'LateAircraftDelay',
       'Month', 'NASDelay', 'Origin', 'SecurityDelay', 'TailNum', 'TaxiIn',
       'TaxiOut', 'UniqueCarrier', 'WeatherDelay', 'Year', 'Code',
       'Description'],
      dtype='object')

In [55]:
cancelled_airlines.groupby(by="UniqueCarrier").sum()[["Cancelled", "Description"]].sort_values(by="Cancelled", ascending=False)

Cancelled                                        Description
UniqueCarrier                                                              
MQ                  8308  American Eagle Airlines Inc.American Eagle Air...
AA                  5844  American Airlines Inc.American Airlines Inc.Am...
OO                  4881  Skywest Airlines Inc.Skywest Airlines Inc.Skyw...
WN                  4462  Southwest Airlines Co.Southwest Airlines Co.So...
DL                  4280  Delta Air Lines Inc.Delta Air Lines Inc.Delta ...
XE                  4019  Expressjet Airlines Inc.Expressjet Airlines In...
OH                  3844  Comair Inc.Comair Inc.Comair Inc.Comair Inc.Co...
UA                  3835  United Air Lines Inc.United Air Lines Inc.Unit...
EV                  3551  Atlantic Southeast AirlinesAtlantic Southeast ...
US                  3123  US Airways Inc. (Merged with America West 9/05...
NW                  2562  Northwest Airlines Inc.Northwest Airlines Inc....
YV                  2405  Mesa Airlines Inc.Mesa Airlines Inc.Mesa Airli...
DH                  1259  Independence AirIndependence AirIndependence A...
AS                  1248  Alaska Airlines Inc.Alaska Airlines Inc.Alaska...
9E                  1208  Pinnacle Airlines Inc.Pinnacle Airlines Inc.Pi...
CO                  1145  Continental Air Lines Inc.Continental Air Line...
FL                   946  AirTran Airways CorporationAirTran Airways Cor...
B6                   726  JetBlue AirwaysJetBlue AirwaysJetBlue AirwaysJ...
HP                   492  America West Airlines Inc. (Merged with US Air...
TZ                   148  ATA Airlines d/b/a ATAATA Airlines d/b/a ATAAT...
F9                   131  Frontier Airlines Inc.Frontier Airlines Inc.Fr...
HA                   108  Hawaiian Airlines Inc.Hawaiian Airlines Inc.Ha...
AQ                    78  Aloha Airlines Inc.Aloha Airlines Inc.Aloha Ai...

### Notes
- Seems like only cancelled flights will have a Cancellation Code -> may want to focus on these flights only if looking at cancelled flights
- Not all cancelled flights have codes through 184011 (without code) vs 58620 (with code)
- Top flights with cancellations are:
  - American Eagle Airlines Inc.
  - American Airlines Inc.
  - Skywest Airlines Inc.
  - Southwest Airlines Co.
  - Delta Air Lines Inc.
  - Expressjet Airlines Inc.
  - Comair Inc.Comair Inc.
  - United Air Lines Inc.
  - Atlantic Southeast Airlines
  - US Airways Inc.

## Checking Diverted Flights

In [4]:
diverted_flights = airline_df[airline_df["Diverted"]==1]
diverted_flights.info()

<class 'pandas.core.frame.DataFrame'>
Index: 22641 entries, 252 to 9881568
Data columns (total 29 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ActualElapsedTime  0 non-null      float64
 1   AirTime            0 non-null      float64
 2   ArrDelay           0 non-null      float64
 3   ArrTime            253 non-null    float64
 4   CRSArrTime         22641 non-null  int64  
 5   CRSDepTime         22641 non-null  int64  
 6   CRSElapsedTime     22396 non-null  float64
 7   CancellationCode   17 non-null     object 
 8   Cancelled          22641 non-null  int64  
 9   CarrierDelay       5397 non-null   float64
 10  DayOfWeek          22641 non-null  int64  
 11  DayofMonth         22641 non-null  int64  
 12  DepDelay           22641 non-null  float64
 13  DepTime            22641 non-null  float64
 14  Dest               22641 non-null  object 
 15  Distance           22578 non-null  float64
 16  Diverted           2264

In [5]:
diverted_flights

ActualElapsedTime  AirTime  ArrDelay  ArrTime  CRSArrTime  \
252                    NaN      NaN       NaN      NaN        2239   
322                    NaN      NaN       NaN      NaN         625   
512                    NaN      NaN       NaN      NaN         830   
558                    NaN      NaN       NaN      NaN        2058   
1189                   NaN      NaN       NaN      NaN        1615   
...                    ...      ...       ...      ...         ...   
9880136                NaN      NaN       NaN      NaN         835   
9880579                NaN      NaN       NaN      NaN          40   
9880782                NaN      NaN       NaN      NaN         646   
9881443                NaN      NaN       NaN      NaN        2052   
9881568                NaN      NaN       NaN      NaN        1129   

         CRSDepTime  CRSElapsedTime CancellationCode  Cancelled  CarrierDelay  \
252            2015           144.0              NaN          0           0.0   
322             550            35.0              NaN          0           NaN   
512            2305           385.0              NaN          0           NaN   
558            1815           103.0              NaN          0           0.0   
1189           1425           170.0              NaN          0           0.0   
...             ...             ...              ...        ...           ...   
9880136         746            49.0              NaN          0           NaN   
9880579          15            25.0              NaN          0           NaN   
9880782         635            71.0              NaN          0           NaN   
9881443        1300           292.0              NaN          0           NaN   
9881568         759           210.0              NaN          0           NaN   

         ...  Month  NASDelay  Origin  SecurityDelay TailNum  TaxiIn  TaxiOut  \
252      ...     10       0.0     EWR            0.0  N21154     0.0     19.0   
322      ...      5       NaN     MLU            NaN     NaN     NaN      NaN   
512      ...      2       NaN     HNL            NaN     NaN     NaN      NaN   
558      ...      3       0.0     ORD            0.0  N933UA     0.0     90.0   
1189     ...      6       0.0     PIT            0.0  N512AE     0.0     10.0   
...      ...    ...       ...     ...            ...     ...     ...      ...   
9880136  ...     11       NaN     SUX            NaN     NaN     NaN      NaN   
9880579  ...     12       NaN     FLL            NaN     NaN     NaN      NaN   
9880782  ...      3       NaN     CVG            NaN  N955UA     0.0     10.0   
9881443  ...      4       NaN     LAX            NaN  N449UA     0.0     41.0   
9881568  ...      7       NaN     PHL            NaN     NaN     NaN      NaN   

         UniqueCarrier  WeatherDelay  Year  
252                 XE           0.0  2006  
322                 DL           NaN  1993  
512                 UA           NaN  1990  
558                 UA           0.0  2006  
1189                MQ           0.0  2007  
...                ...           ...   ...  
9880136             TW           NaN  1989  
9880579             EA           NaN  1990  
9880782             UA           NaN  2000  
9881443             UA           NaN  2000  
9881568             AA           NaN  1994  

[22641 rows x 29 columns]

In [29]:
diverted_flights.to_csv("diverted_flights.csv", index_label=False)

In [56]:
diverted_airlines = pd.merge(diverted_flights, carrier_df, how='left', left_on='UniqueCarrier', right_on='Code')
diverted_airlines.drop(columns=['Code'])
diverted_airlines.groupby(by="UniqueCarrier").sum()[["Diverted", "Description"]].sort_values(by="Diverted", ascending=False)

Diverted                                        Description
UniqueCarrier                                                             
AA                 3404  American Airlines Inc.American Airlines Inc.Am...
US                 2537  US Airways Inc. (Merged with America West 9/05...
UA                 2508  United Air Lines Inc.United Air Lines Inc.Unit...
DL                 2200  Delta Air Lines Inc.Delta Air Lines Inc.Delta ...
WN                 2074  Southwest Airlines Co.Southwest Airlines Co.So...
NW                 2065  Northwest Airlines Inc.Northwest Airlines Inc....
CO                 1842  Continental Air Lines Inc.Continental Air Line...
TW                  855  Trans World Airways LLCTrans World Airways LLC...
AS                  810  Alaska Airlines Inc.Alaska Airlines Inc.Alaska...
XE                  636  Expressjet Airlines Inc.Expressjet Airlines In...
MQ                  575  American Eagle Airlines Inc.American Eagle Air...
EA                  550  Eastern Air Lines Inc.Eastern Air Lines Inc.Ea...
HP                  461  America West Airlines Inc. (Merged with US Air...
OO                  360  Skywest Airlines Inc.Skywest Airlines Inc.Skyw...
EV                  264  Atlantic Southeast AirlinesAtlantic Southeast ...
B6                  249  JetBlue AirwaysJetBlue AirwaysJetBlue AirwaysJ...
PI                  239  Piedmont Aviation Inc.Piedmont Aviation Inc.Pi...
FL                  238  AirTran Airways CorporationAirTran Airways Cor...
OH                  236  Comair Inc.Comair Inc.Comair Inc.Comair Inc.Co...
9E                  137  Pinnacle Airlines Inc.Pinnacle Airlines Inc.Pi...
YV                  128  Mesa Airlines Inc.Mesa Airlines Inc.Mesa Airli...
DH                  116  Independence AirIndependence AirIndependence A...
PA (1)               99  Pan American World Airways (1)Pan American Wor...
F9                   28  Frontier Airlines Inc.Frontier Airlines Inc.Fr...
ML (1)               11  Midway Airlines Inc. (1)Midway Airlines Inc. (...
PS                    7  Pacific Southwest AirlinesPacific Southwest Ai...
TZ                    6  ATA Airlines d/b/a ATAATA Airlines d/b/a ATAAT...
HA                    5  Hawaiian Airlines Inc.Hawaiian Airlines Inc.Ha...
AQ                    1                                Aloha Airlines Inc.

### Notes
- Top Diverted Airlines:
  - American Airlines Inc.
  - Airways Inc.
  - United Air Lines Inc.
  - Delta Air Lines Inc.
  - Southwest Airlines Co.
  - Northwest Airlines Inc.
  - Continental Air Lines Inc.
  - Trans World Airways LLC
  - Alaska Airlines Inc.
  - Expressjet Airlines Inc.

## Checking Delayed Flights

In [30]:
delayed_flight = airline_df[(airline_df['CarrierDelay'] > 0) | 
                 (airline_df['WeatherDelay'] > 0) | 
                 (airline_df['NASDelay'] > 0) | 
                 (airline_df['SecurityDelay'] > 0) | 
                 (airline_df['LateAircraftDelay'] > 0)]
delayed_flight.info()

<class 'pandas.core.frame.DataFrame'>
Index: 679845 entries, 0 to 9882796
Data columns (total 29 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   ActualElapsedTime  679845 non-null  float64
 1   AirTime            679845 non-null  float64
 2   ArrDelay           679845 non-null  float64
 3   ArrTime            679845 non-null  float64
 4   CRSArrTime         679845 non-null  int64  
 5   CRSDepTime         679845 non-null  int64  
 6   CRSElapsedTime     679845 non-null  float64
 7   CancellationCode   0 non-null       object 
 8   Cancelled          679845 non-null  int64  
 9   CarrierDelay       679845 non-null  float64
 10  DayOfWeek          679845 non-null  int64  
 11  DayofMonth         679845 non-null  int64  
 12  DepDelay           679845 non-null  float64
 13  DepTime            679845 non-null  float64
 14  Dest               679845 non-null  object 
 15  Distance           679845 non-null  float64
 16  Divert

In [31]:
delayed_flight.to_csv("delayed_flights.csv", index_label=False)

In [8]:
# delayed_flight = pd.read_csv("delayed_flights.csv")
delayed_flight[delayed_flight.UniqueCarrier == "AA"].Year.value_counts()

Year
2007    14114
2008    13127
2004    11859
2005    11683
2006    11530
2003     6601
Name: count, dtype: int64

In [30]:
aa_delays = pd.DataFrame(delayed_flight[delayed_flight.UniqueCarrier == "AA"].Year.value_counts()).sort_values(by="Year")
aa_delays = aa_delays.rename(columns={"count":"AA_flight_delays"})
aa_delays

AA_flight_delays
Year                  
2003              6601
2004             11859
2005             11683
2006             11530
2007             14114
2008             13127

In [59]:
delayed_airlines = pd.merge(delayed_flight, carrier_df, how='left', left_on='UniqueCarrier', right_on='Code')
delayed_airlines.drop(columns=['Code'])
# count as long as there is a delay
delayed_airlines["total_delay"] = np.where(
    (delayed_airlines['CarrierDelay'] > 0) | 
    (delayed_airlines['WeatherDelay'] > 0) | 
    (delayed_airlines['NASDelay'] > 0) | 
    (delayed_airlines['SecurityDelay'] > 0) | 
    (delayed_airlines['LateAircraftDelay'] > 0), 
    1, 
    0
)
delayed_airlines.groupby(by="UniqueCarrier").sum()[["total_delay", "Description"]].sort_values(by="total_delay", ascending=False)

total_delay                                        Description
UniqueCarrier                                                                
WN                   87706  Southwest Airlines Co.Southwest Airlines Co.So...
AA                   68914  American Airlines Inc.American Airlines Inc.Am...
DL                   53721  Delta Air Lines Inc.Delta Air Lines Inc.Delta ...
MQ                   52838  American Eagle Airlines Inc.American Eagle Air...
UA                   50659  United Air Lines Inc.United Air Lines Inc.Unit...
US                   44132  US Airways Inc. (Merged with America West 9/05...
NW                   43936  Northwest Airlines Inc.Northwest Airlines Inc....
OO                   42260  Skywest Airlines Inc.Skywest Airlines Inc.Skyw...
XE                   40497  Expressjet Airlines Inc.Expressjet Airlines In...
EV                   32850  Atlantic Southeast AirlinesAtlantic Southeast ...
CO                   31296  Continental Air Lines Inc.Continental Air Line...
OH                   25390  Comair Inc.Comair Inc.Comair Inc.Comair Inc.Co...
FL                   22399  AirTran Airways CorporationAirTran Airways Cor...
AS                   16744  Alaska Airlines Inc.Alaska Airlines Inc.Alaska...
YV                   15836  Mesa Airlines Inc.Mesa Airlines Inc.Mesa Airli...
B6                   15485  JetBlue AirwaysJetBlue AirwaysJetBlue AirwaysJ...
DH                    9552  Independence AirIndependence AirIndependence A...
9E                    7991  Pinnacle Airlines Inc.Pinnacle Airlines Inc.Pi...
HP                    7769  America West Airlines Inc. (Merged with US Air...
F9                    5247  Frontier Airlines Inc.Frontier Airlines Inc.Fr...
TZ                    2805  ATA Airlines d/b/a ATAATA Airlines d/b/a ATAAT...
HA                    1350  Hawaiian Airlines Inc.Hawaiian Airlines Inc.Ha...
AQ                     468  Aloha Airlines Inc.Aloha Airlines Inc.Aloha Ai...

### Notes:
- Top delayed flights:
  - Southwest Airlines Co.
  - American Airlines Inc.
  - Delta Air Lines Inc.
  - American Eagle Airlines Inc.
  - United Air Lines Inc.
  - US Airways Inc.
  - Northwest Airlines Inc.
  - Skywest Airlines Inc.
  - Expressjet Airlines Inc.
  - Atlantic Southeast Airlines